lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+/// constructed symbol table to make this fast.
+FModuleLike GrandCentralPass::getEnclosingModule(Value value) {
+  if (auto blockArg = value.dyn_cast<BlockArgument>())
+    return cast<FModuleOp>(blockArg.getOwner()->getParentOp());
 
-  SmallVector<Attribute> newArgAttrs;
-  auto ports = cast<FModuleLike>(op).getPorts();
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    auto port = ports[i];
-    handleRefLike(op, port.annotations, port.type);
-    newArgAttrs.push_back(port.annotations.getArrayAttr());
-  }
+  auto *op = value.getDefiningOp();
+  if (InstanceOp instance = dyn_cast<InstanceOp>(op))
+    return getSymbolTable().lookup<FModuleOp>(
+        instance.moduleNameAttr().getValue());
 
-  op->setAttr("portAnnotations", ArrayAttr::get(op->getContext(), newArgAttrs));
+  return op->getParentOfType<FModuleOp>();
 }
 
-void GrandCentralVisitor::visitDecl(InstanceOp op) {
-
-  // If this instance's underlying module has a "companion" annotation, then
-  // move this onto the actual instance op.
-  AnnotationSet annotations(op.getReferencedModule());
-  if (auto anno = annotations.getAnnotation(
-          "sifive.enterprise.grandcentral.ViewAnnotation")) {
-    auto tpe = anno.getAs<StringAttr>("type");
-    if (!tpe) {
-      op.getReferencedModule()->emitOpError(
-          "contains a ViewAnnotation that does not contain a \"type\" field");
-      failed = true;
-      return;
+/// This method contains the business logic of this pass.
+void GrandCentralPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===- Running Grand Central Views/Interface Pass "
+                             "-----------------------------===\n");
+
+  CircuitOp circuitOp = getOperation();
+
+  // Look at the circuit annotaitons to do two things:
+  //
+  // 1. Determine extraction information (directory and filename).
+  // 2. Populate a worklist of all annotations that encode interfaces.
+  //
+  // Remove annotations encoding interfaces, but leave extraction information as
+  // this may be needed by later passes.
+  SmallVector<Annotation> worklist;
+  bool removalError = false;
+  AnnotationSet::removeAnnotations(circuitOp, [&](Annotation anno) {
+    if (anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType")) {
+      worklist.push_back(anno);
+      return true;
     }
-    if (tpe.getValue() == "companion")
-      op->setAttr("lowerToBind", BoolAttr::get(op.getContext(), true));
-  }
+    if (anno.isClass(
+            "sifive.enterprise.grandcentral.ExtractGrandCentralAnnotation")) {
+      if (maybeExtractInfo.hasValue()) {
+        emitCircuitError("more than one 'ExtractGrandCentralAnnotation' was "
+                         "found, but exactly one must be provided");
+        removalError = true;
+        return false;
+      }
 
-  visitUnhandledDecl(op);
-}
+      auto directory = anno.getMember<StringAttr>("directory"),
+           filename = anno.getMember<StringAttr>("filename");
```

Comment:
:) please split these into two statements. I see this same thing in at least one other place.